### PR TITLE
🪆✈️✈️✈️ react native navigation ✈️✈️✈️🪆

### LINKED ISSUE #9641: No spinner on logout
# Bug Report

## Problem

Currntly it looks like app is freezing for several seconds on logout.
Very visible on slow devices

#### Expected behavior
Spinner until login screen is shown

#### Actual behavior
"Unknown" is shown for a while when log out
logout

### Reproduction

[comment]: # (Describe how we can replicate the bug step by step.)

- Open Status
- Create account
- Logout

### Additional Information

- Status version: nightly 11/12/2019
- Operating System: Android, iOS

### LINKED ISSUE #11556: 🏃‍♀️ Improving app startup time at UX level
# Problem
The time is takes for Status app to be interactive is too much compared to industry standard like WhatsApp or Telegram. In a convoluted experiment on an iPhone SE, with biometrics enabled and many channels joined, it took about 6 seconds for the app to become usable.

A part of the delay is due to the encrypted database - we can't really do anything about this. But a significant chunk is because how we structure screens and components:
[Pre Login]
- After the splash screen goes away, a list of multi accounts show up for a split second
- The user is then redirected to the last used multiaccount (aka login page)
- If the bio-metrics are turned on, the user is then redirected to a biometrics screen automatically

[Post Login]:
- On successful login, we present the user with a loading spinner ⭕️ 
- Then the active chats populate
- Followed by a few seconds delay (which makes me feel that the app has hanged), and finally the new chat messages (ignoring the scroll context)

## Implementation
The implementation is two phased:
1. Change the routing structure to show the login or biometric page directly after splash screen (user can cancel and choose other accounts)
2. Add skeleton screens on chat list page (and maybe inside chats too) instead of a full screen loading indicator, so that the app becomes responsive and user can at least use tabs. 

## Acceptance Criteria
- The bio-metric screen shows up immediately after the splash screen
- The chat screen loading indicator is localised (ie. instead of blocking the entire view, skeletons should be rendered and other parts of the app, like bottom tab bar should be usable).

## Notes
- Instead of skeleton screens on chat list page, we can also show gibberish text like "ejkj787932Fwws" in places where the channel names and messages show up. The loading indicator will say "decrypting" instead of "loading" which make the extra load time justified. 
- The loading skeletons might cascade to other screens to, because if the user can navigate to other tabs before the database is decrypted, many functionalities will break.

### PR DESCRIPTION
**Pull request by flexsurfer**
react native navigation

fixes: #9641
fixes: #11556

This PR is replacing  https://github.com/react-navigation with https://github.com/wix/react-native-navigation

Why?

react-navigation is written with javascript, and uses js thread for actions and animations, also it uses one root for the entire application
react-native-navigation is native, so all elements (buttons, bottom bar .. ) animations and actions are native, js thread isn't used, each screen has its own root, modals, and sheets rendered in a separate container, so they don't affect the main screen

so react-native-navigation should be more performant and feels more native

but because all controls are native we can tune their appearance only with a fixed number of properties, which is less flexible

Also react-native-navigation controls statusBar and navigationBar (android) styles

New deps:
"react-native-navigation": "^7.13.0"

Removed deps:
"@react-navigation/bottom-tabs": "^5.8.0",
"@react-navigation/native": "^5.7.3",
"@react-navigation/stack": "^5.9.0",
"react-native-navigation-bar-color": "^2.0.1"
"react-native-screens": "^2.10.1"

major issues:

- [ ] [ios] dotIndicator is shown on the wrong tab :( https://github.com/wix/react-native-navigation/issues/7146 , currently we show only dotIndicator on ios for chat tab

issues:

- [ ] [ios] screen with autofocus in text input opens with visible lag https://github.com/wix/react-native-navigation/issues/7019
- [x] [android] click on bottomtab doesn't return to the root screen
- [x] [android] bottomtabs no border (elevation has no effect)
- [x] [ios] white background during modal opening animation in dark mode
- [x] hide preview on ios
- [x] [android] statusbar is black (https://wix.github.io/react-native-navigation/api/options-statusBar)
- [x] pressable component doesn't work on android (https://docs.swmansion.com/react-native-gesture-handler/docs/#with-wixreact-native-navigation)
- [x] intro carousel is broken
- [x] reanimated 2.0 throws errors

TODO:
- [x] dark theme
- [x] complex navigations
- [x] screen on focus
- [x] send transaction bottome sheet
- [x] tabs bages

Next:
- [ ] custom topbars

Latest build to try:

| :heavy_check_mark: | 3a80501f | [#45](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-12141/45/) | 2021-06-10 11:41:56 | ~13 min | `android` | [:package:`apk`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-210610-112844-3a8050-pr12141-universal.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-210610-112844-3a8050-pr12141-universal.apk)|
| :heavy_check_mark: | 3a80501f | [#44](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-12141/44/) | 2021-06-10 11:41:58 | ~13 min | `android-e2e` | [:package:`apk`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-210610-112844-3a8050-pr12141-x86.apk) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fstatus-im-prs.ams3.cdn.digitaloceanspaces.com%2FStatusIm-210610-112844-3a8050-pr12141-x86.apk)|
| :heavy_check_mark: | 3a80501f | [#43](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-12141/43/) | 2021-06-10 11:43:24 | ~14 min | `ios` | [:package:`ipa`](https://i.diawi.com/gi9Puo) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fi.diawi.com%2Fgi9Puo)|

### CODE REVIEW

**Comment by jakubgs** [comment]
The error is:
```
Execution failed for task ':app:lintVitalPr'.
> Could not resolve all artifacts for configuration ':react-native-navigation:reactNative63DebugUnitTestRuntimeClasspath'.
   > Could not resolve org.robolectric:robolectric:4.4.
     Required by:
         project :react-native-navigation
      > No cached version of org.robolectric:robolectric:4.4 available for offline mode.
      > No cached version of org.robolectric:robolectric:4.4 available for offline mode.
```
All I did is followed [these instructions](https://github.com/status-im/status-react/tree/develop/nix/deps/gradle#known-issues):
```
echo 'org.robolectric:robolectric:4.4' | go-maven-resolver >> nix/deps/gradle/deps.urls
sort -uo nix/deps/gradle/deps.urls nix/deps/gradle/deps.urls
nix/deps/gradle/generate.sh gen_deps_json
```
But it failed with:
```
Execution failed for task ':app:lintVitalRelease'.
> Could not resolve all artifacts for configuration ':react-native-navigation:reactNative63DebugUnitTestRuntimeClasspath'.
   > Could not resolve org.assertj:assertj-core:3.8.0.
     Required by:
         project :react-native-navigation
      > No cached version of org.assertj:assertj-core:3.8.0 available for offline mode.
      > No cached version of org.assertj:assertj-core:3.8.0 available for offline mode.
```
And then I redid the same seps but for `org.assertj:assertj-core:3.8.0` ... but it failed again with:
```
Execution failed for task ':app:lintVitalRelease'.
> Could not resolve all artifacts for configuration ':react-native-navigation:reactNative63DebugUnitTestRuntimeClasspath'.
   > Could not resolve com.squareup.assertj:assertj-android:1.1.1.
     Required by:
         project :react-native-navigation
      > No cached version of com.squareup.assertj:assertj-android:1.1.1 available for offline mode.
      > No cached version of com.squareup.assertj:assertj-android:1.1.1 available for offline mode.
```
This seems to be a bigger issue.

**Comment by jakubgs** [comment]
So here's a patch that fixes the build: https://gist.github.com/jakubgs/f03d16295a93377a1adb643e8e0c4f99

But the issue with that is that these dependencies disappear when I run `make nix-update-gradle`, so things are still broken.

But at least this patch should unblock you to work on this while I research a bit more.

**Comment by jakubgs** [comment]
This is interesting:
```
[nix-shell:~/repos/status-react]$ nix/deps/gradle/get_deps.sh react-native-navigation | grep org.robolectric:robolectric:4.4

[nix-shell:~/repos/status-react]$ cd android

[nix-shell:~/repos/status-react/android]$ ./gradlew --no-daemon --console plain react-native-navigation:dependencies | grep org.robolectric:robolectric:4.4
+--- org.robolectric:robolectric:4.4
+--- org.robolectric:robolectric:4.4
+--- org.robolectric:robolectric:4.4
+--- org.robolectric:robolectric:4.4
+--- org.robolectric:robolectric:4.4
+--- org.robolectric:robolectric:4.4
+--- org.robolectric:robolectric:4.4 (n)
+--- org.robolectric:robolectric:4.4
```
Looks like our AWK script is not including this dependency in the output from `get_deps.sh`.

**Comment by jakubgs** [comment]
This is the full output from:
```
./gradlew --no-daemon --console plain react-native-navigation:dependencies
```
https://gist.github.com/jakubgs/8e7dd5d226170415e99708463731311c

Our AWK script looks for targets from that command that are `classpath` or other alpha-numeric targets:
https://github.com/status-im/status-react/blob/e5b7fb13a1a3b574c38e5a1cbf39491e0f623046/nix/deps/gradle/gradle_parser.awk#L20

But the target names for `react-native-navigation` that include dependencies like `org.robolectric:robolectric` actually include numbers in their names:
```
 >./gradlew --no-daemon --console plain react-native-navigation:dependencies | \
     awk '/^(classpath|[a-zA-Z0-9]+)( - .*)?$/{ target=$1 } /org.robolectric:robolectric:4.4/{ print target }'
reactNative63DebugUnitTestCompileClasspath
reactNative63DebugUnitTestImplementationDependenciesMetadata
reactNative63DebugUnitTestRuntimeClasspath
reactNative63ReleaseUnitTestCompileClasspath
reactNative63ReleaseUnitTestImplementationDependenciesMetadata
reactNative63ReleaseUnitTestRuntimeClasspath
testImplementation
```
Which means the proper fix would be to add `0-9` into our AWK regex for matching Gradle targets.

**Comment by jakubgs** [comment]
Whoops, pushed the fix to `develop`, but it works, so just rebase on it and run `make nix-update-gradle` and squash out my temporary fix.

**Comment by cammellos** [comment]
a couple of issues I noticed on my side:
1) Opening start a group chat fails : Cannot read property Xh of null
2) Can't go to communities (Tap on communities, click on a community and you get `Not found`)
3) There's a weird flickering on the android top bar when I click on the `+` sign and it opens the `Start new chat...` slider (but to be honest, that's not an issue at all)
4) Communities section has a back arrow (it's hidden in develop)

Those are the ones I found so far,

but the performance difference is astonishing, especially on low powered devices. Amazing work, really impressive!

**Comment by flexsurfer** [comment]
thanks
- Opening start a group chat fails : Cannot read property Xh of null
- Can't go to communities (Tap on communities, click on a community and you get Not found)

fixed

**Comment by Serhy** [comment]
**Issue1**
After account creation/recover - can't login or do any other action inside the app (in one session) because 'Wrong password' all the time.
**Reproduction:**
1) Create an acount 
2) Try to add second account in Wallet OR logout of account and try to login with the same password 

**Issue2**
Wallet balance is empty on the account where ETH/tokens are present. However, transaction history is correct.
**Reproduction**:
Recover account with funds in Mainnet and check Wallet main/account view

File: src/status_im/chat/models.cljs
Comment (shivekkhurana) [inline src/status_im/chat/models.cljs]: Change required?

File: src/status_im/bottom_sheet/core.cljs
Comment (shivekkhurana) [inline src/status_im/bottom_sheet/core.cljs:6]: [SELF NOTE] : Check how this extra effect works.

File: src/status_im/contact/core.cljs
Comment (shivekkhurana) [inline src/status_im/contact/core.cljs:125]: [SELF NOTE] : Verify the existence of this screen.

**Comment by errorists** [comment]
> [android] bottomtabs no border (elevation has no effect)

@flexsurfer can we please fix this with a 1px top border? similar like we have for top bars

**Comment by flexsurfer** [comment]
thanks @Serhy . im unable to replicate Issue1

**Comment by Serhy** [comment]
@flexsurfer thanks! Few more issues with the latest (`7236c3`) build (they are without reproduction mostly, so please let me know if you need more details for any of below)

- [x] **Issue3** Tab back button to put Status on Background -> can’t start Status anymore. only ForceStop helps. To reproduce: from any view you are you now tap hardware Back button so the Status put on back ground and try open Status from the background state again.

- [x] **Issue4** After user is blocked from public chat - when I return back then there is an “Unknown” chat view

- [ ] **Issue5** (Similar as above Issue4) ‘Unknown’ chat opens when I tap ‘Chat’ button from contact profile view

- [x] **Issue6** Can’t open BuyCrypto Dapps from Wallet (`Wallet` -> `Buy crypto` -> `MoonPay`)

- [x] **Issue7** Double tap on Browser Tab button has no effect in respect to return to Browser main tab view (all other tab buttons working OK)

- [x] **Issue8** On iOS dap.ps doesn’t load the content (all fine on Android)

- [ ] **Issue9** There is no three-word signing phrase pop-up when opening the Wallet tab for the first time (however it appears when I try to send a transaction)

- [x] **Issue10** After changed Appearance -> Light/Dark -> all tab views are blank but Chat, and once I open Chats view -> all the rest tabs are “normal” again

**Comment by Serhy** [comment]
- [x] **Issue11** During account recover with Keycard as the key-storage - there is unnecessary ‘Back up seed phrase’ steps requested. (with empty seed phrase value).

**Reproduction**:
1) Open Status
2) `Get started` -> `Access existing keys` -> `Recover with seed phrase` 
3) Enter valid seed phrase -> `Next` -> select `Keycard` as storage 
4) Tap `Begin setup` -> hold keycard to the back of device
5) Create and confirm valid PIN code
**Actual result:** 
Screenshot 2021-06-08 at 16 41 10

(It's all OK when create new account with Keycard storage)

**Comment by churik** [comment]
- [x] **issue 12** can't close other user profile if was scanned from QR (to reproduce: scan chat key qr, try to close view via X)

to continue work with app, needs to be restarted

- [x] **issue 13** error at attempt to scan valid wallet address via universal qr scanner, wallet qr scanner

IMAGE 2021-06-08 17:01:13

- [x] **issue 14** redirect to unknown chat when remove user membership request, delete chat

Steps:
- tap on group invite chat link
- request membership
- ... > Remove
or 
- try to delete public chat

- [x] **issue 15** several back buttons

IMAGE 2021-06-08 17:39:22

**Comment by Serhy** [comment]
- [x] **Issue16**

Similar as Issue3 above. With enabled notifications there is no way to proceed using Status after closing the app, - there is a blank view with Status logo displayed.  
**Reproduction**:
1) Create new account
2) Enable notifications
3) Close Status app from the list of apps(no Logout)
4) Open Status again
**Actual result:** App stays until Force Stop done on the 
Screenshot 2021-06-08 at 19 02 21

**Comment by Serhy** [comment]
@errorists in this PR along performance improvements there is one button replacement happened. And need you opinion whether it's better now of we have to not change it:

- [x] **Issue17** When user recovers account the flow is: Open `Status` -> `Access existing keys` -> `Recover with seed phrase` -> enter valid seed phrase -> tap `Next`. The next screen is the `Your keys have beed successfully recovered` and we always had at the bottom the `Re-encrypt your keys` button which leads to `Chose key Storage' view. Now it's simply `Next` (see below screenshot)

Do you think it makes sense to keep `Next` instead of `Re-encrypt your keys`? 
Screenshot 2021-06-09 at 11 03 51

**Comment by errorists** [comment]
I'm thinking the one on the left, but it's not a strong difference I'm fine with keeping the `Next`. it's only a single button, people will push it regardless how it's labeled.

**Comment by Serhy** [comment]
- [x] **Issue18**

We need to add non-transparent background when navigating between views because navigation doesn't look "smooth":

https://user-images.githubusercontent.com/8749671/121329002-7b9c2f80-c91d-11eb-8f31-c2dd2e50acb0.mov

**Comment by flexsurfer** [comment]
Issue7 Double tap on Browser Tab button has no effect in respect to return to Browser main tab view (all other tab buttons working OK)

as designed, there is no main tab view for browser tab

File: src/status_im/navigation/roots.cljs
Comment (shivekkhurana) [inline src/status_im/navigation/roots.cljs:14]: Align maps maybe?

File: src/status_im/navigation/roots.cljs
Comment (shivekkhurana) [inline src/status_im/navigation/roots.cljs:63]: Dedicated roots map is helpful. Why is having 2 ns for screens a problem?

File: src/status_im/popover/core.cljs
Comment (shivekkhurana) [inline src/status_im/popover/core.cljs:8]: Is this re-factor for now or later?

File: src/status_im/signing/core.cljs
Comment (shivekkhurana) [inline src/status_im/signing/core.cljs:220]: The `#` seems odd here. Is it needed?

File: src/status_im/ui/components/bottom_panel/views.cljs
Comment (shivekkhurana) [inline src/status_im/ui/components/bottom_panel/views.cljs:107]: Is this meant to be merged like this?

File: src/status_im/ui/components/keyboard_avoid_presentation.cljs
Comment (shivekkhurana) [inline src/status_im/ui/components/keyboard_avoid_presentation.cljs:17]: Clever us of `update` with  `+`

File: src/status_im/ui/components/react.cljs
Comment (shivekkhurana) [inline src/status_im/ui/components/react.cljs:233]: Map kv can be aligned.

**Comment by flexsurfer** [comment]
Issue4 After user is blocked from public chat - when I return back then there is an “Unknown” chat view
 Issue5 (Similar as above Issue4) ‘Unknown’ chat opens when I tap ‘Chat’ button from contact profile view
Issue8 On iOS dap.ps doesn’t load the content (all fine on Android)

unable to replicate

**Comment by flexsurfer** [comment]
Issue9 There is no three-word signing phrase pop-up when opening the Wallet tab for the first time (however it appears when I try to send a transaction)

this looks like a feature :)

File: src/status_im/ui/screens/communities/views.cljs
Comment (shivekkhurana) [inline src/status_im/ui/screens/communities/views.cljs:126]: Is this the problem about having to us RN topbar you mentioned on the call yesterday?

File: src/status_im/ui/screens/intro/carousel.cljs
Comment (shivekkhurana) [inline src/status_im/ui/screens/intro/carousel.cljs:1]: [SELF NOTE] Check if this is moved to some other ns

File: src/status_im/ui/screens/multiaccounts/key_storage/views.cljs
Comment (shivekkhurana) [inline src/status_im/ui/screens/multiaccounts/key_storage/views.cljs:109]: [SELF NOTE] Is `:ignore-offset` prop documented in the component?

File: src/status_im/ui/screens/multiaccounts/views.cljs
Comment (shivekkhurana) [inline src/status_im/ui/screens/multiaccounts/views.cljs:38]: How is this `topbar-button` used?

File: src/status_im/ui/screens/screens.cljs
Comment (shivekkhurana) [inline src/status_im/ui/screens/screens.cljs:231]: There are multiple TODOs in this NS. Do you intend to address these before merge?

File: src/status_im/ui/screens/wallet/account/views.cljs
Comment (shivekkhurana) [inline src/status_im/ui/screens/wallet/account/views.cljs:213]: Should this be replaced with native topbar?

File: src/status_im/ui/screens/wallet/add_new/views.cljs
Comment (shivekkhurana) [inline src/status_im/ui/screens/wallet/add_new/views.cljs:124]: [SELF NOTE] Do we ever use `:ignore-offset false`?

File: src/status_im/ui/screens/wallet/recipient/views.cljs
Comment (shivekkhurana) [inline src/status_im/ui/screens/wallet/recipient/views.cljs]: Rogue comment.

**Review by shivekkhurana — APPROVED** [review]
This is an extraordinary piece of work. It is the second time since I've been here that Andrey has pulled a feat like this (earlier, he refactored the events system).

I have left many comments. The ones marked with `[SELF NOTE]` are for me to check and resolve. There are a few small issues that I've pointed out.

File: src/status_im/ui/screens/intro/carousel.cljs
Comment (shivekkhurana) [reply src/status_im/ui/screens/intro/carousel.cljs:1]: It is.

File: src/status_im/chat/models.cljs
Comment (flexsurfer) [reply src/status_im/chat/models.cljs]: Outdated

File: src/status_im/bottom_sheet/core.cljs
Comment (flexsurfer) [reply src/status_im/bottom_sheet/core.cljs:6]: now we need to show overlay, before bottom-sheet were mounted from the beggining

File: src/status_im/contact/core.cljs
Comment (flexsurfer) [reply src/status_im/contact/core.cljs:125]: it was a first screen in the create-group-chat stack, now we don't have a stack, so we just open a screen

File: src/status_im/navigation/roots.cljs
Comment (flexsurfer) [reply src/status_im/navigation/roots.cljs:14]: sure

File: src/status_im/popover/core.cljs
Comment (flexsurfer) [reply src/status_im/popover/core.cljs:8]: later

File: src/status_im/signing/core.cljs
Comment (flexsurfer) [reply src/status_im/signing/core.cljs:220]: yes, thats how macro works

File: src/status_im/ui/components/bottom_panel/views.cljs
Comment (flexsurfer) [reply src/status_im/ui/components/bottom_panel/views.cljs:107]: yes

File: src/status_im/ui/screens/multiaccounts/views.cljs
Comment (flexsurfer) [reply src/status_im/ui/screens/multiaccounts/views.cljs:38]: `(.registerNavigationButtonPressedListener
   (.events Navigation)`
called from here

File: src/status_im/ui/screens/screens.cljs
Comment (flexsurfer) [reply src/status_im/ui/screens/screens.cljs:231]: no, for later

File: src/status_im/ui/screens/wallet/add_new/views.cljs
Comment (flexsurfer) [reply src/status_im/ui/screens/wallet/add_new/views.cljs:124]: false is default

**Comment by Serhy** [comment]
- [x] **Issue19** On iOS when creating group chat bottom navigation control panel has a gap between keyboard/tabs 

https://user-images.githubusercontent.com/8749671/121341945-d8054c00-c929-11eb-976c-5e35d57a6d3c.mov

**Comment by flexsurfer** [comment]
thanks @churik and @Serhy all should be fixed except keycard

**Comment by flexsurfer** [comment]
@Serhy keycard fixed

**Comment by Serhy** [comment]
- [x] **Issue20**: Three-words passphrase does not appear when sending transaction from Wallet, making impossible to send transaction (shadowed background and nothing happens when I tap anywhere on screen).

**Reproduction**:
1) Recover account with funds (e.g. for Ropsten - `record sphere illegal section blame swing wreck estate celery same excite summer`)
2) Switch to Ropsten network
3) Try to send transaction from Wallet to any address
Actual result of `Sending...` view:
Screenshot 2021-06-10 at 19 18 11

**Comment by flexsurfer** [comment]
@Serhy fixed

**Comment by Serhy** [comment]
Those failed e2e tests will be fixed separately as they are mostly e2e specific (on real devices it looks good).
Ultra great job here, @flexsurfer thank you! Navigation between screens even with naked eye visible how much faster it became.

**Pull request merged**
